File: src/Dime/InvoiceBundle/Entity/InvoiceItem.php
<?php
/**
 * Author: Till Wegmüller
 * Date: 9/24/14
 * Dime
 */

namespace Dime\InvoiceBundle\Entity;

use Dime\TimetrackerBundle\Entity\Activity;
use Dime\TimetrackerBundle\Entity\Entity;
use Dime\TimetrackerBundle\Model\DimeEntityInterface;
use Doctrine\ORM\Mapping as ORM;
use JMS\Serializer\Annotation as JMS;
use Knp\JsonSchemaBundle\Annotations as Json;
use Money\Money;

/**
 * Class InvoiceItem
 * @package Dime\InvoiceBundle\Entity
 * @ORM\Entity(repositoryClass="Dime\InvoiceBundle\Entity\InvoiceItemRepository")
 * @ORM\Table(name="invoice_items")
 * @Json\Schema("invoiceItems")
 */
class InvoiceItem extends Entity implements DimeEntityInterface
{
    /**
     * @var Invoice
     * @ORM\ManyToOne(targetEntity="Dime\InvoiceBundle\Entity\Invoice", inversedBy="items")
     * @JMS\MaxDepth(1)
     */
    protected $invoice;

    /**
     * @var string
     * @ORM\Column(type="string")
     */
    protected $name;

    /**
     * @var int
     * @ORM\Column(name="rate_value", type="money", nullable=true)
     * @JMS\SerializedName("rateValue")
     * @JMS\Type(name="Money")
     */
    protected $rateValue;

    /**
     * @var string
     * @ORM\Column(type="string", nullable=true)
     * @JMS\SerializedName("rateUnit")
     */
    protected $rateUnit;

    /**
     * @ORM\Column(type="decimal", scale=3, precision=10, nullable=true)
     */
    protected $vat;

    /**
     * @var string
     * @ORM\Column(type="string", nullable=true)
     */
    protected $amount;

    /**
     * @var Activity
     * @ORM\ManyToOne(targetEntity="Dime\TimetrackerBundle\Entity\Activity", )
     */
    protected $activity;

    /**
     * @JMS\SerializedName("order")
     * @ORM\Column(name="order_no", type="decimal", nullable=true)
     */
    protected $order;

    /**
     * @param Activity $activity
     *
     * @return $this
     */
    public function setFromActivity(Activity $activity)
    {
        $this->name         = $activity->getService()->getName();
        $this->rateValue    = $activity->getRateValue();
        $this->amount = $activity->getValue();
        $this->rateUnit     = $activity->getRateUnit();
        $this->vat          = $activity->getVat();
        $this->activity     = $activity;
        return $this;
    }

    /**
     * @JMS\VirtualProperty
     * @JMS\SerializedName("calculatedVAT")
     * @JMS\Type(name="Money")
     * @return Money
     */
    public function getCalculatedVAT()
    {
        if ($this->rateValue instanceof Money && is_numeric($this->amount) && is_numeric($this->vat)) {
            return $this->rateValue->multiply((float)$this->amount)->multiply((float)$this->vat);
        } else {
            return Money::CHF(0);
        }
    }

    public function getCalculatedTotal()
    {
        if ($this->rateValue instanceof Money && is_numeric($this->amount)) {
            return $this->rateValue->multiply((float)$this->amount);
        } else {
            return Money::CHF(0);
        }
    }

    /**
     * @JMS\VirtualProperty()
     * @JMS\SerializedName("total")
     * @JMS\Type(name="Money")
     * @return Money
     */
    public function getTotal()
    {
        $total = $this->getRateValue();
        if ($total === null) {
            return null;
        }
        if ($this->getAmount() !== null) {
            $total = $total->multiply($this->getAmount());
        }
        $vat = $this->getCalculatedVAT();
        if ($vat instanceof Money) {
            $total = $total->add($vat);
        }
        return $total;
    }

    /**
     * @JMS\VirtualProperty()
     * @JMS\SerializedName("totalWithoutVAT")
     * @JMS\Type(name="Money")
     * @return Money
     */
    public function getTotalWithoutVAT()
    {
        $total = $this->getRateValue();
        if ($total === null) {
            return null;
        }
        if ($this->getAmount() !== null) {
            $total = $total->multiply($this->getAmount());
        }
        return $total;
    }

    /**
     * @return Invoice
     */
    public function getInvoice()
    {
        return $this->invoice;
    }

    /**
     * @param Invoice $invoice
     *
     * @return $this
     */
    public function setInvoice($invoice)
    {
        $this->invoice = $invoice;
        return $this;
    }

    /**
     * @return string
     */
    public function getRateUnit()
    {
        return $this->rateUnit;
    }

    /**
     * @param string $rateUnit
     *
     * @return $this
     */
    public function setRateUnit($rateUnit)
    {
        $this->rateUnit = $rateUnit;
        return $this;
    }

    /**
     * @return Money
     */
    public function getRateValue()
    {
        return $this->rateValue;
    }

    /**
     * @param Money $rateValue
     *
     * @return $this
     */
    public function setRateValue($rateValue)
    {
        $this->rateValue = $rateValue;
        return $this;
    }

    /**
     * @return string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @param string $name
     *
     * @return $this
     */
    public function setName($name)
    {
        $this->name = $name;
        return $this;
    }

    /**
     * @return float
     */
    public function getAmount()
    {
        return floatval($this->amount);
    }

    /**
     * @param float $amount
     *
     * @return $this
     */
    public function setAmount($amount)
    {
        $this->amount = $amount;
        return $this;
    }

    /**
     * @return mixed
     */
    public function getVat()
    {
        return $this->vat ? $this->vat : 0;
    }

    /**
     * @param mixed $vat
     *
     * @return $this
     */
    public function setVat($vat)
    {
        $this->vat = $vat;
        return $this;
    }

    /**
     * @return Activity
     */
    public function getActivity()
    {
        return $this->activity;
    }

    /**
     * @param Activity $activity
     *
     * @return $this
     */
    public function setActivity($activity)
    {
        $this->activity = $activity;
        return $this;
    }

    /**
     * @return mixed
     */
    public function getOrder()
    {
        return $this->order;
    }

    /**
     * @param mixed $order
     */
    public function setOrder($order)
    {
        $this->order = $order;
    }
}

File: src/Dime/FrontendBundle/Resources/public/lib/src/model/entities/InvoiceItem.dart
import '../Entity.dart';

class InvoiceItem extends Entity {
  @override
  init({Map<String, dynamic> params: const {}}) {
    if (!params.containsKey('name')) {
      params['name'] = 'New Item';
    }
    if (params.containsKey('invoice')) {
      params['invoice'] = new Invoice()..id = params['invoice'];
    }
    super.init(params: params);
  }

  InvoiceItem();

  InvoiceItem.clone(InvoiceItem original) : super.clone(original) {
    this.order = original.order;
    this.name = original.name;
    this.amount = original.amount;
    this.rateValue = original.rateValue;
    this.rateUnit = original.rateUnit;
    this.activity = original.activity;
    this.vat = original.vat;
    this.invoice = original.invoice;
    addFieldstoUpdate(['order', 'name', 'amount', 'rateValue', 'rateUnit', 'activity', 'vat', 'invoice']);
  }

  InvoiceItem.fromMap(Map<String, dynamic> map) : super.fromMap(map);

  @override
  InvoiceItem newObj() {
    return new InvoiceItem();
  }

  @override
  dynamic Get(String property) {
    var val = super.Get(property);
    if (val == null) {
      switch (property) {
        case 'order':
          return this.order;
        case 'rateValue':
          return this.rateValue;
        case 'rateUnit':
          return this.rateUnit;
        case 'amount':
          return this.amount;
        case 'activity':
          return this.activity;
        case 'vat':
          return this.vat;
        case 'invoice':
          return this.invoice;
        case 'calculatedVAT':
          return this.calculatedVAT;
        case 'total':
          return this.total;
        default:
          break;
      }
    }
    return val;
  }

  @override
  void Set(String property, dynamic value) {
    switch (property) {
      case 'order':
        this.order = value as int;
        break;
      case 'rateValue':
        this.rateValue = value as String;
        break;
      case 'rateUnit':
        this.rateUnit = value as String;
        break;
      case 'amount':
        this.amount = value;
        break;
      case 'activity':
        this.activity = value is Activity ? value : new Activity.fromMap(value as Map<String, dynamic>);
        break;
      case 'vat':
        this.vat = value as double;
        break;
      case 'invoice':
        this.invoice = value is Invoice ? value : new Invoice.fromMap(value as Map<String, dynamic>);
        break;
      case 'calculatedVAT':
        this.calculatedVAT = value as String;
        break;
      case 'total':
        this.total = value as String;
        break;
      default:
        super.Set(property, value);
        break;
    }
  }

  static List<InvoiceItem> listFromMap(List<Map<String, dynamic>> content) {
    List<InvoiceItem> array = new List<InvoiceItem>();
    for (var element in content) {
      array.add(new InvoiceItem.fromMap(element));
    }
    return array;
  }

  @override
  String type = 'invoiceitems';
  int order;
  Invoice invoice;
  String rateValue;
  String rateUnit;
  String calculatedVAT;
  dynamic amount;
  String total;
  Activity activity;
  double vat;
}
